### PR TITLE
Add audit event for Kubernetes authenticator client certificate injection

### PR DESCRIPTION
**Pull request by micahlee**
### What does this PR do?
- _What's changed? Why were these changes made?_

This PR makes two key changes:

- It factors the authentication audit event into separate events with the same `authn`
  action, but distinct `operation` values that. There are now individual
  operations for `authenticate`, `login`, and `validate-status`.

- It adds an audit event, `k8s-inject-client-certificate`, for the client certificate injection
  step in the Kubernetes authenticator workflow.

- _How should the reviewer approach this PR, especially if manual tests are required?_

The history is setup to allow reviewing the operations refactor separate from the new audit event for `k8s-inject-client-cert`.

### Example Audit Events

**Authenticate (Successful)**:

image

```
{"subject@43868":{"role":"demo:host:conjur/authn-k8s/demo/apps/test-app-demo/service_account/oc-test-app-secretless"},"auth@43868":{"service":"demo:webservice:conjur/authn-k8s/demo","authenticator":"authn-k8s"},"action@43868":{"result":"success","operation":"authenticate"},"PROGRAM":"conjur","PID":"2d46cbc0-30b2-4cae-a931-b130e01f2285","MSGID":"authn","MESSAGE":"demo:host:conjur/authn-k8s/demo/apps/test-app-demo/service_account/oc-test-app-secretless successfully authenticated with authenticator authn-k8s service demo:webservice:conjur/authn-k8s/demo","LEVEL":"info","ISODATE":"2020-05-14T18:54:19.786+00:00","FACILITY":"authpriv"}
```

**Login (Successful)**:

image

```
{"subject@43868":{"role":"demo:user:admin"},"auth@43868":{"authenticator":"authn"},"action@43868":{"result":"success","operation":"login"},"PROGRAM":"conjur","PID":"895e066e-69bb-435d-bbc5-82d6e725ae53","MSGID":"authn","MESSAGE":"demo:user:admin successfully logged in with authenticator authn","LEVEL":"info","ISODATE":"2020-05-14T14:35:32.312+00:00","FACILITY":"authpriv"}
```

**Login (Unsuccessful)**:

image

```
{"subject@43868":{"role":"demo:user:admin"},"auth@43868":{"authenticator":"authn"},"action@43868":{"result":"failure","operation":"login"},"PROGRAM":"conjur","PID":"4028353f-3f60-4e3c-b4c1-38bb9268af47","MSGID":"authn","MESSAGE":"demo:user:admin failed to login with authenticator authn: CONJ00002E Invalid credentials","LEVEL":"warning","ISODATE":"2020-05-14T14:36:51.145+00:00","FACILITY":"authpriv"}
```

**Inject Client Certificate (Successful)**:

image

```
{"subject@43868":{"role":"demo:host:conjur/authn-k8s/demo/apps/test-app-demo/service_account/oc-test-app-secretless"},"auth@43868":{"service":"demo:webservice:conjur/authn-k8s/demo","authenticator":"authn-k8s"},"action@43868":{"result":"success","operation":"k8s-inject-client-cert"},"PROGRAM":"conjur","PID":"c081a5e1-256e-4f12-8b9d-66ce5b40efc1","MSGID":"authn","MESSAGE":"demo:host:conjur/authn-k8s/demo/apps/test-app-demo/service_account/oc-test-app-secretless successfully injected client certificate with authenticator authn-k8s service demo:webservice:conjur/authn-k8s/demo","LEVEL":"info","ISODATE":"2020-05-14T18:54:19.727+00:00","FACILITY":"authpriv"}
```

### What ticket does this PR close?
Connected to #1538 

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] This PR does not require updating any documentation, or
- [x] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs

#### Follow-up Issues
- [x] [DAP Wiki audit docs are updated](https://github.com/cyberark/dap-wiki/pull/24)
- [x] [Appliance tests run cleanly with these changes](https://jenkins.conjur.net/job/conjurinc--appliance/job/1538-audit-inject-cert/)

### CODE REVIEW

File: app/domain/authentication/audit_log.rb [hunk 1]
```diff
@@ -1,17 +1,35 @@
 # frozen_string_literal: true
 
 module Authentication
-  class AuditLog
-    def self.record_authn_event(role:, webservice_id:, authenticator_name:,
-                                success:, message: nil)
+
+  AuditLog = CommandClass.new(
```
Comment (jonahx) [inline app/domain/authentication/audit_log.rb:5]: `CommandClass` names should be verb phrases.  Not sure what's most appropriate here, maybe something like `WriteAuditEvent`?

File: app/domain/authentication/audit_log.rb [hunk 2]
```diff
@@ -1,17 +1,35 @@
...
-        authenticator_name: authenticator_name,
-        service: Resource[webservice_id],
-        success: success,
-        error_message: message
-      ).log_to Audit.logger
+        authenticator_name: @authenticator_input.authenticator_name,
+        service: @resource_cls[webservice_id],
+        success: @success,
+        error_message: @message
+      ).log_to ::Audit.logger
```
Comment (jonahx) [inline app/domain/authentication/audit_log.rb:22]: `::Audit.logger` looks like a dependency.

**Comment by micahlee** [comment]
@jonahx , this should be ready for review again when you have a moment. I also took the opportunity to split of the 1 large initial commit to better factor out the cleanup stuff from actual changes.

**Comment by micahlee** [comment]
@jonahx , sorry I wasn't able to finish this back up as quickly as I'd hoped, but... I've taken care of the low hanging fruit that was obvious to me.

Let me know if you want to work on any more in the context of this PR, or whether you're comfortable cutting this here, and working on the next iteration in the context of the larger audit work.

**Pull request merged**